File: src/main/java/mod/lucky/network/SpawnPacket.java
package mod.lucky.network;

import mod.lucky.Lucky;
import net.minecraft.client.Minecraft;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.network.PacketBuffer;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.fml.network.NetworkEvent;

import java.util.function.Supplier;

public class SpawnPacket {
    private CompoundNBT entityTag;

    public SpawnPacket() {}

    public SpawnPacket(Entity entity) {
        this.entityTag = entity.serializeNBT();
    }

    public static SpawnPacket decode(PacketBuffer buf) {
        SpawnPacket msg = new SpawnPacket();
        msg.entityTag = buf.readCompoundTag();
        return msg;
    }

    public static void encode(SpawnPacket msg, PacketBuffer buf) {
        buf.writeCompoundTag(msg.entityTag);
    }

    public static class Handler {
        @OnlyIn(Dist.CLIENT)
        private static void spawn(SpawnPacket msg) {
            ClientWorld world = Minecraft.getInstance().world;

            Entity entity = EntityType.loadEntityAndExecute(msg.entityTag, world, e -> {
                world.addEntity(e.getEntityId(), e);
                return e;
            });

            if (entity == null) {
                Lucky.error(null, "Invalid entity spawning on the client side: "
                    + msg.entityTag.toString());
                return;
            }
        }

        public static void handle(SpawnPacket msg, Supplier<NetworkEvent.Context> context) {
            context.get().enqueueWork(() -> {
                spawn(msg);
            });
            context.get().setPacketHandled(true);
        }
    }
}

File: forge/src/main/kotlin/game/SpawnPacket.kt
package mod.lucky.forge.game

import mod.lucky.common.gameAPI
import mod.lucky.forge.CompoundTag
import mod.lucky.forge.ForgeLuckyRegistry
import mod.lucky.forge.OnlyInClient
import net.minecraft.client.world.ClientWorld
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityType
import net.minecraft.network.*
import net.minecraftforge.fml.network.NetworkDirection
import net.minecraftforge.fml.network.NetworkInstance
import net.minecraftforge.fml.network.simple.SimpleChannel
import java.nio.channels.NetworkChannel

/*
data class SpawnPacket(
    var entityNBT: CompoundTag
) { companion object }

fun SpawnPacket.Companion.fromEntity(entity: Entity): SpawnPacket {
    val nbt = CompoundTag()
    entity.saveAsPassenger(nbt)
    return SpawnPacket(nbt)
}

fun SpawnPacket.toPacket(): IPacket<*> {
    return NetworkDirection.PLAY_TO_CLIENT.buildPacket<IPacket<*>>(org.apache.commons.lang3.tuple.Pair.of(encode(), 0), ForgeLuckyRegistry.spawnPacketId).getThis()
}

@OnlyInClient
fun SpawnPacket.execute(world: ClientWorld) {
    val entity = EntityType.loadEntityWithPassengers(entityNBT, world) { e ->
        world.addEntity(e.id, e)
        e
    }

    if (entity == null) {
        gameAPI.logError("Error spawning client entity with NBT '$entityNBT'")
        return
    }
}

fun SpawnPacket.Companion.decode(buf: PacketByteBuf): SpawnPacket? {
    val nbt = buf.readNbt()
    if (nbt == null) {
        gameAPI.logError("Error decoding entity spawn packet")
        return null
    }
    return SpawnPacket(nbt)
}

fun SpawnPacket.encode(): PacketBuffer {
    val buf = PacketBuffer.create()
    buf.writeNbt(entityNBT)
    return buf
}

 */